### PR TITLE
BLE: improve permission and signing strategy

### PR DESCRIPTION
**Pull request by pan-**
This PR defines new security requirements for `GattCharacteristics`; previously a single security requirements was applied to the characteristic and used for all operations. This prevented common use cases such as when a characteristic doesn't require permission to be read bu require some permission to be written. 

Starting with this patch a characteristic defines three independent permission requirements: 
* Read: applied for all read operations
* Write: applied for write without response, signed write, write or queued writes.
* Update: It defines the write permission of the CCCD and the security required to send an update.

Similarly descriptors also posses independent permissions: read and write. 

The type `SecurityManager::SecurityMode_t` was not modeling requirement appropriately as it doesn't even model security more correctly. Therefore the type `att_security_requirement_t` has been introduced to properly define what a security requirement is.
* `att_security_requirement_t::NONE`: the operation do not have security requirements.
* `att_security_requirement_t::UNAUTHENTICATED`: the operation requires the use of an unauthenticated or an authenticated key.
* `att_security_requirement_t::AUTHENTICATED`: the operation requires the use of an authenticated key.
* `att_security_requirement_t::SC_AUTHENTICATED`: the operation requires the use of an authenticated key shared over a secure connection.

It maps well with LE security mode 1 (regular GATT operations) and LE security mode 2 (signed operation): 

|                                              | Security Mode 1 | Security Mode 2 |
|----------------------------------------------|-----------------|-----------------|
| att_security_requirement_t::NONE             |     level 1     |        N/A        |
| att_security_requirement_t::UNAUTHENTICATED  |     level 2  - encrypted    |     level 1 - signed      |
| att_security_requirement_t::AUTHENTICATED    |     level 3 - encrypted with authenticated key     |     level 2 - signed with authenticated CSRK    |
| att_security_requirement_t::SC_AUTHENTICATED |     level 4 - encrypted with authenticated key from SC pairing     |        N/A        |

Cordio and Nordic implementations have been updated to take into account these new permissions and the relationship between permissions and properties have been strengthen. As an example it is a programming fault to define a characteristic that accepts signed write with permissions not being UNAUTHENTICATED or AUTHENTICATED.

### CODE REVIEW

**Comment by pan-** [comment]
@paul-szczepanek-arm Could you review ?

File: features/FEATURE_BLE/ble/BLETypes.h [hunk 1]
```diff
@@ -460,6 +461,74 @@ struct random_address_type_t : SafeEnum<random_address_type_t, uint8_t> {
         SafeEnum<random_address_type_t, uint8_t>(value) { }
 };
 
+/**
+ * Security requirement that can be attached to an attribute operation.
+ */
+struct att_security_requirement_t : SafeEnum<att_security_requirement_t, uint8_t> {
+    /**
```
[the remark this answers is not in the document]
Comment (pan-) [reply features/FEATURE_BLE/ble/BLETypes.h]: The use case you present can be solve by an appropriate uses of properties: 

If you want a write to succeed for authenticated peer whether encryption is present or not then set the properties to write and write and signedWrite and the the security requirement to AUTHENTICATED.

If you do not want signedWrite but wants authentication then don't set the signedWriteProperty property and keep the security requirement to AUTHENTICATE; it is as simple as that. 

One interesting aspects of the security modes is the automatic conversion of security mode 2 into security mode 1 once encryption is enabled: signed write cannot be used over an encrypted and the standard mandate to use regular write. 

If authentication is required to achieve a signed write it makes sense that authentication is also a requirement over an encrypted link. Same for non authenticated CSRK; if signed writes are accepted without authentication then it makes sense that regular write accepted are accepted if the link is not encrypted. 

In my opinion, a differentiation between the two would create an API easy to misuses as the key requirement for the write operation may be different depending on the context.

File: features/FEATURE_BLE/ble/GattCharacteristic.h [hunk 2]
```diff
@@ -1371,13 +1406,19 @@ class GattCharacteristic {
...
         _descriptors(descriptors),
         _descriptorCount(numDescriptors),
-        enabledReadAuthorization(false),
-        enabledWriteAuthorization(false),
         readAuthorizationCallback(),
-        writeAuthorizationCallback() {
+        writeAuthorizationCallback(),
+        _update_security(SecurityRequirement_t::NONE) {
+        _valueAttribute.allowWrite(isWritable(_properties));
+        _valueAttribute.allowWrite(isReadable(_properties));
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/ble/GattCharacteristic.h]: allow read

File: features/FEATURE_BLE/ble/GattCharacteristic.h [hunk 3]
```diff
@@ -1371,13 +1406,19 @@ class GattCharacteristic {
...
+
+        if (_properties & BLE_GATT_CHAR_PROPERTIES_AUTHENTICATED_SIGNED_WRITES) {
+            _valueAttribute.setWriteSecurityRequirement(
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/ble/GattCharacteristic.h]: needs a comment explaining why this translation was chosen

Comment (pan-) [reply features/FEATURE_BLE/ble/GattCharacteristic.h]: Signed writes requires at least Security Mode 2 level 1 which requires at least an unauthenticated CSRK so if signed writes are set then the lowest security requirement is unauthenticated.

File: features/FEATURE_BLE/ble/pal/MemorySecurityDb.h [hunk 4]
```diff
@@ -276,8 +276,16 @@ class MemorySecurityDb : public SecurityDb {
         for (size_t i = 0; i < MAX_ENTRIES; i++) {
             if (_entries[i].state == ENTRY_FREE) {
                 continue;
-            } else if (peer_address == _entries[i].peer_identity.identity_address
-                && _entries[i].flags.peer_address_is_public == peer_address_public) {
+            } else if (
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/ble/pal/MemorySecurityDb.h]: looks like you already fixed this in a different PR?

Comment (pan-) [reply features/FEATURE_BLE/ble/pal/MemorySecurityDb.h]: Not sure about that; we talked about it for sure. If it was fixed then why is there a diff and a new condition ?

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 5]
```diff
@@ -413,6 +429,19 @@ ble_error_t GenericSecurityManager::setLinkEncryption(
             return requestAuthentication(connection);
         }
 
+    } else if (encryption == link_encryption_t::ENCRYPTED_WITH_SC_AND_MITM) {
+
+        if (cb->ltk_mitm_protected &&
+            cb->secure_connections_paired && !
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: && !
that is confusing, I think these should be on the next line

**Comment by paul-szczepanek-arm** [comment]
I don't get it why we need separate security for update, can't we just use the read/write security reqs for that?

**Comment by pan-** [comment]
Being allow to read or write doesn't mean your authorized to register and receive update. Am I missing something here ?

**Comment by paul-szczepanek-arm** [comment]
No, I mean security permissions. You don't need to store the security, just a bit saying you are also allowed to receive updates.

**Comment by pan-** [comment]
That's for the GattServer. According to the specification CCCD's shall support unauthenticated read and the write requirements are defined by the application. The middleware also needs to know what the security requirements are to send updates.

**Comment by pan-** [comment]
@paul-szczepanek-arm I've addressed your comments; please review.

@donatieng Is there anything left to address on your side ?

**Pull request merged**